### PR TITLE
Application Analytics

### PR DESCRIPTION
**Pull request by eugenesk24**
### Description
App Analytics skeleton to be merged into main

### Issues Resolved
Start Application Analytics

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Application.kt [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+ *   "traceGroups": [
+ *       "Payment.auto",
+ *       "Users.admin",
+ *       "Purchase.source"
+ *   ],
+ *   "availabilityLevels": [
+ *       {
+ *           "label": "Unavailable",
+ *           "color": "#D36086",
+ *           "condition": "when errorRate() is above or equal to 2%",
```
Comment (ps48) [inline opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Application.kt:43]: should the condition start with "where" instead of "when"

File: dashboards-observability/public/components/trace_analytics/components/traces/__tests__/traces.test.tsx [hunk 2]
```diff
@@ -33,6 +33,8 @@ describe('Traces component', () => {
         endTime="now"
         setEndTime={setEndTime}
         indicesExist={false}
+        hasTitle={true}
+        page="traces"
```
Comment (ps48) [inline dashboards-observability/public/components/trace_analytics/components/traces/__tests__/traces.test.tsx:37]: you might want to add tests with both type of inputs for pages.

File: dashboards-observability/public/components/explorer/home.tsx [hunk 3]
```diff
@@ -55,6 +56,12 @@ import { DeletePanelModal } from '../custom_panels/helpers/modal_containers';
 import { CUSTOM_PANELS_API_PREFIX } from '../../../common/constants/custom_panels';
 import { getSampleDataModal } from '../common/helpers/add_sample_modal';
 
+const pageStyles: CSS.Properties = {
+  float: 'left',
+  width: '100%',
+  maxWidth: '1130px',
+};
+
```
Comment (ps48) [inline dashboards-observability/public/components/explorer/home.tsx:64]: I think this is already present in home.scss file. These lines might be redundant.

File: dashboards-observability/public/components/trace_analytics/components/dashboard/dashboard.tsx [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 import dateMath from '@elastic/datemath';
 import { EuiFlexGroup, EuiFlexItem, EuiSpacer, EuiTitle } from '@elastic/eui';
+import _ from 'lodash';
```
Comment (ps48) [inline dashboards-observability/public/components/trace_analytics/components/dashboard/dashboard.tsx:8]: is this import used anywhere in file?

File: dashboards-observability/public/components/application_analytics/home.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+          render={() => 
+            <CreateApp
+            dslService={dslService}
+            {...commonProps}
+            />
+          }
+        />
+        <Route
+          exact
+          path={'/application_analytics/:id+'}
```
Comment (ps48) [inline dashboards-observability/public/components/application_analytics/home.tsx:131]: Question: Just curious what's the plus after id for?

File: dashboards-observability/public/components/application_analytics/home.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+};
+
+const dateString = new Date().toISOString();
+
+const dummyApplication: ApplicationType[] = [{
+  name: "Cool Application", 
+  id: "id", 
+  composition: "Payment, user_db",
+  currentAvailability: "Available",
+  availabilityMetrics: "Error rate: 0.80%, Throughput: 0.94%, Latency: 600ms",
```
Comment (ps48) [inline dashboards-observability/public/components/application_analytics/home.tsx:51]: Are we pushing dummy data to main branch? If App Analytics is fully functional we can remove this.

**Review by ps48 — COMMENTED** [review]
It would be great if you can add jest & cypress tests for App Analytics (preferred in this PR or can do it later as well). You should add more jest tests for modules changed in trace and explorer.

File: dashboards-observability/public/components/application_analytics/components/trace_config.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        newFilters.splice(i, 1, newFilter);
+        setFilters(newFilters);
+        return;
+      }
+    }
+  }
+
+  const clearTraces = () => {
+    const withoutTraces = filters.filter((f) => f.field !== 'traceGroup')
+    filters.map((f) => {console.log(f.field !== 'traceGroup')})
```
Comment (joshuali925) [inline dashboards-observability/public/components/application_analytics/components/trace_config.tsx:151]: remove this?

File: dashboards-observability/public/components/application_analytics/components/trace_config.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+          onChange={onTraceChange}
+          onCreateOption={onCreateTrace}
+          isClearable={false}
+          data-test-subj="traceGroupsComboBox"
+        />
+      </EuiFormRow>
+      <EuiSpacer />
+      <DashboardTable
+        items={traceItems}
+        filters={[]}
```
Comment (joshuali925) [inline dashboards-observability/public/components/application_analytics/components/trace_config.tsx:193]: maybe add a comment on why this is `[]`?

File: dashboards-observability/public/components/application_analytics/home.tsx (hunk 6 above)
Comment (joshuali925) [reply dashboards-observability/public/components/application_analytics/home.tsx:51]: if this has to be here, maybe put it in a separate mock data directory and then import the constant?

File: dashboards-observability/public/components/application_analytics/components/app_table.tsx [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+import CSS from 'csstype';
+import _ from 'lodash';
+import React, { ReactElement, useEffect, useState } from 'react';
+import { AppAnalyticsComponentDeps, ApplicationType } from '../home';
+
+const pageStyles: CSS.Properties = {
+  float: 'left',
+  width: '100%',
+  maxWidth: '1130px',
+};
```
Comment (joshuali925) [inline dashboards-observability/public/components/application_analytics/components/app_table.tsx:37]: this appeared multiple times (also in notebooks), maybe define in constant files?

File: dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx [hunk 10]
```diff
@@ -58,7 +59,7 @@ const getType = (field: string): string => {
     startTime: 'date_nanos',
   };
   const type = _.get(typeMapping, field, 'keyword');
-  return typeof type === 'string' ? type : null;
+  return type;
```
Comment (joshuali925) [inline dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx:62]: why this change?

File: dashboards-observability/public/components/trace_analytics/components/services/services.tsx [hunk 11]
```diff
@@ -13,24 +13,43 @@ import { filtersToDsl } from '../common/helper_functions';
...
+  ] : [
+    {
         text: 'Trace analytics',
         href: '#/trace_analytics/home',
       },
       {
-        text: 'Services',
-        href: '#/trace_analytics/services',
+        text: 'Dashboards',
+        href: '#/trace_analytics/home',
```
Comment (joshuali925) [inline dashboards-observability/public/components/trace_analytics/components/services/services.tsx:45]: this shouldn't change?

File: dashboards-observability/public/components/trace_analytics/components/traces/traces.tsx [hunk 12]
```diff
@@ -12,24 +12,43 @@ import { filtersToDsl } from '../common/helper_functions';
...
+  ] : [
+    {
         text: 'Trace analytics',
         href: '#/trace_analytics/home',
       },
       {
-        text: 'Traces',
-        href: '#/trace_analytics/traces',
+        text: 'Dashboards',
+        href: '#/trace_analytics/home',
```
Comment (joshuali925) [inline dashboards-observability/public/components/trace_analytics/components/traces/traces.tsx:44]: this also shouldn't change

File: opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Application.kt [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+ *           "order": "0",
+ *       }
+ *   ],
+ * }
+ * }</pre>
+ */
+
+internal data class Application(
```
Comment (joshuali925) [inline opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Application.kt:51]: could you also add tests for this (example https://github.com/opensearch-project/trace-analytics/blob/main/opensearch-observability/src/test/kotlin/org/opensearch/observability/model/TimestampTests.kt)? maybe in another PR

File: dashboards-observability/public/components/application_analytics/components/app_table.tsx (hunk 9 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/components/app_table.tsx:37]: Is it okay to change the pageStyle width to 1130 and panelStyle width to 1110?? This aligns with the rest of the plugins.
Screen Shot 2021-12-13 at 11 37 19 AM

Comment (joshuali925) [reply dashboards-observability/public/components/application_analytics/components/app_table.tsx:37]: sure if it doesn't look broken

Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/components/app_table.tsx]: Didn't change the one on notebooks

File: dashboards-observability/public/components/application_analytics/home.tsx (hunk 6 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/home.tsx:51]: App Analytics is not fully functional yet so we have dummy data for now.

File: dashboards-observability/public/components/application_analytics/home.tsx (hunk 5 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/home.tsx:131]: If the id includes a "/", the + makes sure it is not seen as the end of the id and gets the full id including the "/" and everything after

File: dashboards-observability/public/components/application_analytics/home.tsx (hunk 6 above)
Comment (ps48) [reply dashboards-observability/public/components/application_analytics/home.tsx:51]: @joshuali925 If the feature is still in development should we have a separate branch for it?

File: dashboards-observability/public/components/explorer/home.tsx (hunk 3 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/explorer/home.tsx]: You're right! Thanks!

File: dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx (hunk 10 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx:62]: I thought line 61 wasn't necessary because the typeof type is always a string by definition

File: dashboards-observability/public/components/application_analytics/home.tsx (hunk 6 above)
Comment (joshuali925) [reply dashboards-observability/public/components/application_analytics/home.tsx:51]: following opensearch standards we can merge to main, but i think it's better to keep in a separate branch during development so any external user won't get a broken UI using default branch

File: dashboards-observability/public/components/trace_analytics/components/dashboard/dashboard.tsx (hunk 4 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/trace_analytics/components/dashboard/dashboard.tsx:8]: Looks like it's not! I'll remove it.

File: dashboards-observability/public/components/trace_analytics/components/services/services.tsx (hunk 11 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/trace_analytics/components/services/services.tsx:45]: Oh yea, don't know why that's there. Good catch!

File: opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Application.kt (hunk 1 above)
Comment (eugenesk24) [reply opensearch-observability/src/main/kotlin/org/opensearch/observability/model/Application.kt:43]: The structure of availability level conditions are currently being discussed/changed so this will be updated later

File: dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx (hunk 10 above)
Comment (joshuali925) [reply dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx:62]: not exactly, if `field` is `attributes`, `type` is an object. But I forgot what was the use case here, you can keep this change and revert if anything breaks

File: dashboards-observability/public/components/application_analytics/home.tsx (hunk 6 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/home.tsx:51]: Just create branch application-analytics and change to merge into there!

File: dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx (hunk 10 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/trace_analytics/components/common/filters/filter_helpers.tsx:62]: Ohh that makes sense. No I'll put it back.

File: dashboards-observability/public/components/trace_analytics/components/traces/__tests__/traces.test.tsx (hunk 2 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/trace_analytics/components/traces/__tests__/traces.test.tsx:37]: Right, right. I'll add tests in a separate PR later.

**Review by ps48 — APPROVED** [review]

**Review by ps48 — APPROVED** [review]

File: dashboards-observability/public/components/application_analytics/components/application.tsx [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+  const getService = () => {
+    return (
+      <Services {...props} page="app" appId="id" appName="Cool Application" />
+    );
+  };
+
+  const getTrace = () => {
+    return (
+      <>
+      <Traces {...props} page="app" appId="id" appName="Cool Application" />
```
Comment (mengweieric) [inline dashboards-observability/public/components/application_analytics/components/application.tsx:114]: Minor: indent issue

File: dashboards-observability/public/components/application_analytics/components/application.tsx [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+        savedObjects={savedObjects}
+        timestampUtils={timestampUtils}
+        http={http} 
+        history={history} 
+        notifications={notifications} 
+        setToast={setToast}
+        // App Analytics will not be saving queries on Log Events
+        savedObjectId={''} 
+        getExistingEmptyTab={getExistingEmptyTab}
+        />
```
Comment (mengweieric) [inline dashboards-observability/public/components/application_analytics/components/application.tsx:139]: Minor: Also indent issue here and some other places.

File: dashboards-observability/public/components/application_analytics/components/application.tsx [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+          }
+        )
+    ];
+  };
+  
+
+  const memorizedAppAnalyticsTabs = useMemo(() => {
+    return getAppAnalyticsTabs();
+  },
+  []);
```
Comment (mengweieric) [inline dashboards-observability/public/components/application_analytics/components/application.tsx:220]: You don't need any dependencies?

**Review by mengweieric — APPROVED** [review]
Left few minor comments.

File: dashboards-observability/public/components/application_analytics/components/application.tsx (hunk 14 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/components/application.tsx:114]: Will fix in next PR

File: dashboards-observability/public/components/application_analytics/components/application.tsx (hunk 15 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/components/application.tsx:139]: Will fix in next PR

File: dashboards-observability/public/components/application_analytics/components/application.tsx (hunk 16 above)
Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/components/application.tsx:220]: With no dependencies, it will be called once on mount which is okay

Comment (eugenesk24) [reply dashboards-observability/public/components/application_analytics/components/application.tsx:220]: But I can also just change it to be used without useMemo, will change in next PR

Comment (ps48) [reply dashboards-observability/public/components/application_analytics/components/application.tsx:220]: Ideally everywhere we use useMemo, it should have array filled up with dependencies used by its functions/components: https://reactjs.org/docs/hooks-reference.html#usememo

> Conceptually though, that’s what they represent: every value referenced inside the function should also appear in the dependencies array.
